Apps/Src/ReadCarCAN.c

```diff
@@ -67,98 +93,145 @@ static inline void chargingEnable(void) {
     // if nothing is precharging and we need to turn the array on, turn it on
     shouldRestartArray = !restartingArray && (Contactors_Get(ARRAY_CONTACTOR)==OFF) && (Contactors_Get(ARRAY_PRECHARGE)==OFF);
 
-    // start precharge for array 
+    // wait for precharge for array 
     if(shouldRestartArray){
 
-         if (Contactors_Set(ARRAY_PRECHARGE, ON) == SUCCESS) {    // only run through precharge sequence if we successfully start precharge
-            restartingArray = true;
+            // Wait to make sure precharge is finished and then restart array
+            OSTmrStart(&prechargeDlyTimer, &err);
             
-            //Turn on the array restart thread
-            OSTaskCreate(
-                &arrayTCB,
-                "Array Restarter",
-                &ArrayRestart,
-                NULL,
-                TASK_READ_CAR_CAN_PRIO,
-                arraySTK,
-                DEFAULT_STACK_SIZE/10,
-                DEFAULT_STACK_SIZE,
-                0,
-                1,
-                NULL,
-                OS_OPT_TASK_STK_CLR,
-                &err
-            );
-            assertOSError(OS_READ_CAN_LOC,err);   
         }
-    }
+    
 
     OSMutexPost(&arrayRestartMutex,
                 OS_OPT_NONE,
                 &err);
     assertOSError(OS_READ_CAN_LOC,err);
 }
 
+/**
+ * @brief Callback function for the precharge delay timer. Turns off precharge and restarts the array.
+ * @param p_tmr pointer to the timer that calls this function, passed by timer
+ * @param p_arg pointer to the argument passed by timer
+ * 
+*/
+static void arrayRestart(void *p_tmr, void *p_arg){
+
+    if(!regenEnable){    // If regen enable has been disabled during precharge, we don't want to turn on the main contactor immediately after
+        restartingArray = false;
+    }
+    else {
+       
+            Contactors_Set(ARRAY_CONTACTOR, ON, SET_CONTACTORS_NON_BLOCKING);
+    }
+
+        //Lights_Set(A_CNCTR, ON); Don't know where this function is. Does it even exist?
```

Comment:
It no longer exists, you should probably replace it with a call to `UpdateDisplay_SetArray(true)`.